src/Bicep.Core/Configuration/ConfigurationManager.cs

```diff
@@ -1,101 +1,153 @@
+        private readonly ConcurrentDictionary<Uri, ConfigLookupResult> configLookupCache = new();
         private readonly IFileSystem fileSystem;
 
         public ConfigurationManager(IFileSystem fileSystem)
         {
             this.fileSystem = fileSystem;
         }
 
-        public RootConfiguration GetBuiltInConfiguration() => BuiltInConfigurationLazy.Value;
-
         public RootConfiguration GetConfiguration(Uri sourceFileUri)
         {
-            var configurationPath = DiscoverConfigurationFile(fileSystem.Path.GetDirectoryName(sourceFileUri.LocalPath));
+            var (config, diagnosticBuilders) = GetConfigurationFromCache(sourceFileUri);
+            return diagnosticBuilders.Count > 0
+                ? new(config.Cloud, config.ModuleAliases, config.Analyzers, config.ConfigurationPath, diagnosticBuilders)
+                : config;
+        }
 
-            if (configurationPath is not null)
+        public (RootConfiguration, List<DiagnosticBuilder.DiagnosticBuilderDelegate>) GetConfigurationFromCache(Uri sourceFileUri)
+        {
+            List<DiagnosticBuilder.DiagnosticBuilderDelegate> diagnostics = new();
+            var loadErrorEncountered = false;
+
+            var (configFileUri, lookupDiagnostic) = configLookupCache.GetOrAdd(sourceFileUri, LookupConfiguration);
+            if (lookupDiagnostic is not null)
             {
-                try
-                {
-                    using var stream = fileSystem.FileStream.Create(configurationPath, FileMode.Open, FileAccess.Read);
-                    var element = BuiltInConfigurationElement.Merge(JsonElementFactory.CreateElement(stream));
+                diagnostics.Add(lookupDiagnostic);
+            }
 
-                    return RootConfiguration.Bind(element, configurationPath);
-                }
-                catch (JsonException exception)
+            if (configFileUri is not null)
+            {
+                var (config, loadError) = configCache.GetOrAdd(configFileUri, LoadConfiguration);
+                if (loadError is not null)
                 {
-                    throw new ConfigurationException($"Failed to parse the contents of the Bicep configuration file \"{configurationPath}\" as valid JSON: \"{exception.Message}\".");
+                    loadErrorEncountered = true;
+                    diagnostics.Add(loadError);
                 }
-                catch (Exception exception) when (exception is IOException or UnauthorizedAccessException or SecurityException)
+
+                if (config is not null)
                 {
-                    throw new ConfigurationException($"Could not load the Bicep configuration file \"{configurationPath}\": \"{exception.Message}\".");
+                    return (config, diagnostics);
                 }
             }
 
-            return GetBuiltInConfiguration();
+            return (GetDefaultConfiguration(loadErrorEncountered), diagnostics);
```

Comment:
Are there scenarios where we could have >1 diagnostics for configuration? Is it helpful to show both, or should we just show the first?